### ISSUE TITLE
nbgv prepare-release should support signing commits

### ISSUE THREAD

**Issue by egil** [dotnet/Nerdbank.GitVersioning#594]
If using `nbgv prepare-release` with a repository that has git signing enabled, the command should issue use the GPG key configured in the GIT repository.

The alternative is to offer a switch that causes `nbgv prepare-release` to not commit, but just update the branches. That covers the hard part, then it is pretty easy to scrip the commits and merges that follows that.

**Why:** if using a GitHub repo that requires signed commits, then the `prepare-release` wont work. 

**Workaround:** I think the best work around for now is to use `git reset` to undo the commit but keep the changes, and the commit using `git commit` to signed commits.

**Comment by KalleOlaviNiemitalo** [comment]
Would `git commit --amend` work and be simpler than resetting first?

**Comment by egil** [comment]
@KalleOlaviNiemitalo I am not sure. On the mainline branch there are two commits, the merge commit and the "Set version to next x.y-preview" commit. This works, is non-interactive (doesnt require an editor), but would love a simpler workaround for sure.

**Comment by AArnott** [comment]
It looks like libgit2sharp supports signed commits, so this should be possible once we update `nbgv` to activate that behavior.
https://github.com/libgit2/libgit2sharp/issues/1101

**Comment by KalleOlaviNiemitalo** [comment]
AFAICT, libgit2sharp has methods for creating commit objects with signatures with them, but not for creating the signatures or accessing the GnuPG keystore. So `nbgv` would have to find and run `gpg` itself.

**Comment by egil** [comment]
I'm thinking that the easiest fix is to just make the changes, but not commit, if `--no-commit` is passed to the command.

**Comment by AArnott** [comment]
@egil The change that you're thinking of is just a one-liner change -- insufficient to justify even the existence of the CLI tool, IMO.
The tool does more than just change that line. It commits, branches, then switches back and commits again. At least, it does this when not _already_ on the release branch. Given this, a `--no-commit` option makes less sense because it would be incompatible with a core workflow.

@KalleOlaviNiemitalo That's a keen observation. Seeking out other executables to execute may complicate matters. How would we find `gpg`? Can we assume it's on the PATH?

**Comment by egil** [comment]
@AArnott I get what you are saying, but to me the part of updating the files is not trivial, since it involves reading the version.json and applying the correct rules to increment the version numbers. Doing two commits and a merge in the right order is however pretty simple, and I would think, a good separation of concerns.

I think the simplest solution would be to check if there is a `git` command line tool set up in the environment you are running in. If there is, use that. It will trigger the signing process if the user has set it up to sign commits. It will also work in my scenario on in Github Actions as there are packages there that enable signing to happen as long as you use the `git commit...` command line tool.

Then you can fall back to the `libgit2sharp` packages if no `git` command line tools are detected.

**Comment by KalleOlaviNiemitalo** [comment]
If a user is running `nbgv` from PowerShell on Windows, I think `gpg` is unlikely to be on the PATH, even though `git` may be.

Git has **gpg.program** and related configuration variables. Perhaps it is enough if `nbgv` first checks those and then searches the PATH.

Would `nbgv` have to support both "openpgp" and "x509" in **gpg.format**? I imagine that would be annoying to implement and test when `git` already has an implementation.

**Comment by egil** [comment]
@KalleOlaviNiemitalo if we just use whatever whatever `git` command line tool is in the path of the user, then that is responsible for integrating with whatever type of signing the user has configured. No need for `nbgv` to be involved in that.

**Comment by KalleOlaviNiemitalo** [comment]
@egil that's correct. However it seems a bit unfortunate if nbgv needs a fourth GitContext implementation, now that it already has LibGit2Context, ManagedGitContext, and NoGitContext.

**Comment by AArnott** [comment]
The git engine property that we now have occurred to me as well. But there's no way the git CLI can replace the in-proc solutions we have now for calculating version height, so conflating them may be misleading. Also only `nbgv` is impacted by this, so I'm thinking a `--use-git-cli` switch to the `prepare-release` command may be the most appropriate.

I'm not a fan of interacting with the git CLI. Doing so means we have to check exit codes, possibly parse stdout/stderr, etc. We would need to consider using the "plumbing" vs. the "porcelain" for a more stable integration with it. 
But I certainly do not want to dive into the gpg vs x509 and all the more complicated alternatives, so invoking `git` and assuming it is on the path may be the simplest way.

There's no way I'll have time for this in the forseeable future, so it'll only happen if it comes in by PR.

**Comment by egil** [comment]
> Also only nbgv is impacted by this, so I'm thinking a `--use-git-cli` switch to the prepare-release command may be the most appropriate.
> ...
> I'm not a fan of interacting with the git CLI.

I agree, it is not simple, there are all sorts of corner cases you need to deal with then doing that.

Since you are not a fan of the `--no-commit` variant, another completely different approach would be to add a new command to NBGV called `get-next-version` or something similar? Then those like me who has this special need can use that to get the next (unstable) version number to bump the mainline version no with (using set-version command) and I should also then be able to branch of mainline and create the release branch myself and set the version to a non-preview variant there using the set-version command.

**Comment by AArnott** [comment]
Sure, I think a `get-next-version` command makes sense.

**Comment by egil** [comment]
Should I close this issue and create a new one dedicated to `get-next-version`?

**Comment by AArnott** [comment]
Let's leave this active a little longer. I like signing all my commits and I simply hadn't noticed that `nbgv prepare-release` didn't sign them, but using that command locally on a dev box shouldn't preclude pushing commits in a repo configured to reject unsigned commits.
But as I don't know how long it will be before I get to this again, and we don't have great options at the moment, feel free to open a new issue to track `get-next-version`.
